cmd/tiller-proxy/handler.go

```diff
@@ -46,78 +79,142 @@ func isAlreadyExists(err error) bool {
 	return strings.Contains(err.Error(), "is still in use") || strings.Contains(err.Error(), "already exists")
 }
 
-func deployRelease(w http.ResponseWriter, req *http.Request, params Params) {
+func isForbidden(err error) bool {
+	return strings.Contains(err.Error(), "Unauthorized")
+}
+
+func errorCode(err error) int {
+	errCode := http.StatusInternalServerError
+	if isAlreadyExists(err) {
+		errCode = http.StatusConflict
+	} else if isNotFound(err) {
+		errCode = http.StatusNotFound
+	} else if isForbidden(err) {
+		errCode = http.StatusForbidden
+	}
+	return errCode
+}
+
+func deployRelease(w http.ResponseWriter, req *http.Request, params Params, user *auth.UserAuth) {
 	log.Printf("Creating/updating Helm Release")
 	defer req.Body.Close()
 	body, err := ioutil.ReadAll(req.Body)
 	if err != nil {
-		response.NewErrorResponse(http.StatusInternalServerError, err.Error()).Write(w)
+		response.NewErrorResponse(errorCode(err), err.Error()).Write(w)
+		return
+	}
+	chartDetails, err := chartUtils.ParseDetails(body)
+	if err != nil {
+		response.NewErrorResponse(http.StatusUnprocessableEntity, err.Error()).Write(w)
 		return
 	}
+	ch, err := chartUtils.GetChart(chartDetails, kubeClient, netClient, chartutil.LoadArchive)
+	if err != nil {
+		response.NewErrorResponse(errorCode(err), err.Error()).Write(w)
+		return
+	}
+	action := "create"
+	if req.Method == "PUT" && params["releaseName"] != "" {
```

Comment:
why not use a different handler for PUT?

Reply:
it has just a couple of lines different but I guess it's better to have different handlers for readability